DOC: add usage section

hagenw opened the pull request:
Closes #5.

This adds a first basic usage section to the documentation, covering:

* using `audinterface.Process`
* using `audinterface.Feature`
* building a feature extraction object with `audinterface.Feature` and `audobject.Object`

Building the documentation will fail as I get the following warning due to storing and loading the feature extraction object:

```
WARNING: Cell printed to stderr:
/home/audeering.local/hwierstorf/.envs/audinterface/lib/python3.8/site-packages/audobject/core/utils.py:51: RuntimeWarning: Could not determine a version for module '__main__'.
  warnings.warn(
```

Any idea how to avoid this?

frankenjoe commented:
> Any idea how to avoid this?

did you try to set `__version__ = '1.0.0' ` at the top the file?

hagenw commented:
> > Any idea how to avoid this?
> 
> did you try to set `__version__ = '1.0.0' ` at the top the file?

Thanks, this works indeed.

hagenw commented on docs/usage.rst:
```diff
@@ -0,0 +1,156 @@
+    index = db['emotion'].index
+
+
+DSP algorithm returning a series
```
I call it here "DSP algorithm" as my original goal was to also add a section for an example using a "Machine learning model". But I'm not sure if this is needed and if we have a small one that we use here.

frankenjoe replied:
I am not sure if really need that, but if we want to, we can create a dummy model with `audonnx.testing`.

frankenjoe replied:
Ok, then I suggest we also choose a more generic name for the section, maybe: "Create a processing interface"

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,162 @@
+The only prerequisite is
+the algorithm provides a callable
+that takes at least the signal
+as a :class:`np.ndarray`
```
```suggestion
as a :class:`numpy.ndarray`
```

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,162 @@
+When using a DSP algorithm as feature extractor,
+it is recommended to use :class:`audinterface.Feature`,
+which returns results as a :class:`pd.DataFrame`
+and assigns names to the single features.
```
Another advantage is that we can set a window and hop duration in case of features with dynamic length. I think we should highlight by adding a second example.

hagenw replied:
I added such an example under the new subsection "Create a framewise feature extractor".
But maybe my implementation of calculating the mean on a framewise basis is slightly clumsy ;)

![image](https://user-images.githubusercontent.com/173624/172867156-942bc33a-426f-4d4f-a4cf-236fefb5c698.png)

frankenjoe replied:
Maybe:

```python
def features(signal, sampling_rate, win_dur, hop_dur):
    win_dur = int(win_dur * sampling_rate)
    hop_dur = int(hop_dur * sampling_rate)
    starts = np.arange(0, signal.shape[1] - win_dur, hop_dur)
    mean = [[signal[:, s:s+win_dur].mean() for s in starts]]
    return mean
```

frankenjoe replied:
Or we switch to an example where we interface an external library, e.g.:

```python
win_dur = 0.02
hop_dur = 0.01
n_mfcc = 13
interface = audinterface.Feature(
    [f'mfcc-{idx}' for idx in range(n_mfcc)],
    process_func=features,
    process_func_args={
        'win_dur': win_dur,
        'hop_dur': hop_dur,
        'n_mfcc': n_mfcc,
    },
    hop_dur=hop_dur,
    win_dur=win_dur,
)
df = interface.process_files(db.files[:10])
df
```
```
                                                                                                      mfcc-0  ...    mfcc-12
file                                               start                  end                                 ...           
/media/jwagner/Data/audb/emodb/1.2.0/fe182b91/w... 0 days 00:00:00        0 days 00:00:00.020000 -611.993286  ...   1.151396
                                                   0 days 00:00:00.010000 0 days 00:00:00.030000 -668.175842  ...  14.068544
                                                   0 days 00:00:00.020000 0 days 00:00:00.040000 -664.612793  ...   7.949759
                                                   0 days 00:00:00.030000 0 days 00:00:00.050000 -667.714722  ...  12.957479
                                                   0 days 00:00:00.040000 0 days 00:00:00.060000 -669.367432  ...   4.396849
...                                                                                                      ...  ...        ...
/media/jwagner/Data/audb/emodb/1.2.0/fe182b91/w... 0 days 00:00:01.650000 0 days 00:00:01.670000 -619.806335  ...  11.462805
                                                   0 days 00:00:01.660000 0 days 00:00:01.680000 -622.713928  ...  13.530191
                                                   0 days 00:00:01.670000 0 days 00:00:01.690000 -624.977356  ...  12.032556
                                                   0 days 00:00:01.680000 0 days 00:00:01.700000 -621.033264  ...  11.019306
                                                   0 days 00:00:01.690000 0 days 00:00:01.710000 -622.040344  ...  11.202772

[1743 rows x 13 columns]
```

frankenjoe replied:
I think it's general a good idea to cover this kind of use-case.

hagenw replied:
Good idea, I switched to `librosa.feature.mfcc`

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,162 @@
+    df
+
+
+Creating a feature extraction object
```
Maybe: Create a serializable feature extractor?

hagenw replied:
Changed to "Create a serializable feature extractor"

frankenjoe replied:
Yes, makes sense.

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,162 @@
+
+.. jupyter-execute::
+
+    fex.to_yaml('mean-std.yml')
```
```suggestion
    fex.to_yaml('mean-std.yaml')
```

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,162 @@
+.. jupyter-execute::
+
+    fex.to_yaml('mean-std.yml')
+    fex2 = audobject.from_yaml('mean-std.yml')
```
```suggestion
    fex2 = audobject.from_yaml('mean-std.yaml')
```

we usually use "yaml" (in `audonnx` we even force it), so I think it's better to always stick to `.yaml`.

frankenjoe commented:
Good to see we will finally have a usage section!

hagenw commented:
At the moment every subsection starts with "Create a"

![image](https://user-images.githubusercontent.com/173624/172868039-8774d7a9-e362-43f9-8e56-15709ea22bd2.png)

should I maybe change to

![image](https://user-images.githubusercontent.com/173624/172868225-e551c769-a569-4aa2-b7d4-cb4dea770e35.png)

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,209 @@
+Create a feature extractor
+--------------------------
+
+When using an algorithm as feature extractor,
```
Maybe we should be a little bit more generic here. E.g.: When the result of the processing function has multiple dimensions it is recommended to use :class:`audinterface.Feature`, which returns a :class:`pandas.DataFrame`
with named dimensions.

hagenw replied:
Changed accordingly.

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,209 @@
+    import os
+
+    media = ['wav/03a01Fa.wav', 'wav/03a01Nc.wav', 'wav/03a01Wa.wav']
+    db = audb.load('emodb', version='1.2.0', media=media, verbose=False)
```
I'd recommend we set `full_path=False` and then use `process_index(..., root=db.root)` to avoid showing the full path in the output.

hagenw replied:
Good to know, maybe we should then also extend the docstring and be more precise there.
I thought when using `root=` it will also add this to the path in the returned index.

But if not, it is indeed much more elegant than my current solution.

hagenw replied:
> it is indeed much more elegant than my current solution.

I just realized, this refers more to what I did in
https://github.com/audeering/audinterface/pull/48,
where I used `audformat.utils.map_file_path()` to remove the absolute path in the output.

Here, in the usage documentation I think it makes more sense to stay with `full_path=True` as this is the behavior the user would most likely need. The resulting dataframes will look ok in the HTML version of the docs, as the full path is usually shorter on the Github runner as it is when building locally.

frankenjoe replied:
ok

hagenw commented:
> Should we also add an example for `Segment`?

I added one, but it was not so easy to find a processing function that returns segments.
For now I ended up using https://auditok.readthedocs.io/

hagenw commented:
> At the moment every subsection starts with "Create a"
> 
> ![image](https://user-images.githubusercontent.com/173624/172868039-8774d7a9-e362-43f9-8e56-15709ea22bd2.png)
> 
> should I maybe change to
> 
> ![image](https://user-images.githubusercontent.com/173624/172868225-e551c769-a569-4aa2-b7d4-cb4dea770e35.png)

I adjusted the headings accordingly, and also changed the last word always to "interface"

frankenjoe commented:
> For now I ended up using https://auditok.readthedocs.io/

Interesting, they even offer real-time detection. Could this also be a candidate benchmark against?

hagenw commented:
> > For now I ended up using https://auditok.readthedocs.io/
> 
> Interesting, they even offer real-time detection. Could this also be a candidate benchmark against?

Maybe, but at least according to this it performs not as good as others: https://github.com/wiseman/py-webrtcvad/issues/68

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,257 @@
+        )
+        return index
+
+    interafce = audinterface.Segment(process_func=segments)
```
```suggestion
    interface = audinterface.Segment(process_func=segments)
```

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,257 @@
+        return index
+
+    interafce = audinterface.Segment(process_func=segments)
+    idx = interafce.process_file(files[0])
```
```suggestion
    idx = interface.process_file(files[0])
```

frankenjoe replied:
I think I would also rename to `index`, with `idx` I associate an integer.

frankenjoe replied:
Ah wait, we already have a variable `index`.

frankenjoe replied:
Actually, is there a specific reason that you assign the result and not just show the output directly (applies also to the other examples).

hagenw replied:
The reason behind is that the user can more easily see what type the method returns.
If we just show the HTML formated output I'm afraid that it is not that easy to differentiate between Series, DataFrame, and Index

frankenjoe replied:
ok

hagenw replied:
And yes, I used `idx` instead of `index` to not overwrite the already assigned `index` variable.

We could call it `segmented_index`, but I was also not convinced to add an additional word as we also don't do it for the other examples.

frankenjoe replied:
Yes, I guess there is no perfect solution, so let's stick with `idx`.

frankenjoe commented on docs/usage.rst:
```diff
@@ -0,0 +1,257 @@
+
+Segmentation interface
+----------------------
+
```
Like with the other sections I think we should provide a short introduction here.

hagenw replied:
Oh, totally forgot that.

I now added

![image](https://user-images.githubusercontent.com/173624/173046418-999f9d08-d30d-4ded-8ad8-a29ebdc27ab1.png)

frankenjoe replied:
Cool, thanks. No more comments, so I think we are good to go.
